### PR TITLE
Add specs for blockCtx ops

### PR DESCRIPTION
**Pull request by bchyl**

### CODE REVIEW

File: src/zkevm_specs/evm/execution/block_ctx.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+from ...util import FQ
+
+
+def blockctx(instruction: Instruction):
+    opcode = instruction.opcode_lookup(True)
+
+    res_op = instruction.curr.execution_state.responsible_opcode()[0]
+    if isinstance(res_op, int):
+        res_op_int = int(res_op)
+    instruction.constrain_equal(opcode, FQ(res_op_int))
```
Comment (ed255) [inline src/zkevm_specs/evm/execution/block_ctx.py:14]: Is this really needed? In `Instruction.step_state_transition_to_new_context` we're doing `self.responsible_opcode_lookup(opcode)` which I think checks the same thing.
The other opcodes are not doing this constraint check.

File: src/zkevm_specs/util/param.py [hunk 2]
```diff
@@ -1,5 +1,5 @@
 # Maximun number of bytes with composition value that doesn't wrap around the field
-MAX_N_BYTES = 31
+MAX_N_BYTES = 32
```
Comment (ed255) [inline src/zkevm_specs/util/param.py:2]: There's no guarantee that 32 bytes don't wrap around the field, so this value should be kept at `31`.

File: src/zkevm_specs/evm/execution/block_ctx.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
+from ...util.param import N_BYTES_ACCOUNT_ADDRESS, N_BYTES_U64
+from ..instruction import Instruction, Transition
+from ..table import BlockContextFieldTag
+from ..opcode import Opcode
+from ...util import FQ
+
+
+def blockctx(instruction: Instruction):
```
Comment (ed255) [inline src/zkevm_specs/evm/execution/block_ctx.py:8]: I have one question about this function.  Currently in the circuits we have one gadget per execution state.  And some execution states handle multiple opcodes.
Will `blockctx` be a single gadget?
If it's a single gadget, I wonder if there's an advantage if we define a BlockContext execution state that handles all the block context opcodes in a single execution state :thinking: Maybe we would reduce the number of constraints a bit?

**Review by ed255 — CHANGES_REQUESTED** [review]

**Comment by ed255** [comment]
Thanks for submitting this PR!  I think merging the logic of all block context opcodes is a nice improvement :)

File: specs/opcode/41COINBASE_45GASLIMIT_48BASEFEE.md [hunk 4]
```diff
@@ -0,0 +1,60 @@
+# block context op codes
```
[the remark this answers is not in the document]
Comment (bchyl) [reply specs/opcode/41COINBASE_45GASLIMIT_48BASEFEE.md:1]: This is supposed to be on range from `41COINBASE` to `45GASLIMIT` and `48 BASEFEE` as trade-offs,  while the md file name as `41COINBASE_42TIMESTAMP_43NUMBER_44DIFFICULTY_45GASLIMIT_48BASEFEE.md` seems too long.
Maybe we can name it as `41COINBASE-45GASLIMIT_48BASEFEE.md`?

File: src/zkevm_specs/evm/execution/block_ctx.py [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+    opcode = instruction.opcode_lookup(True)
+
+    res_op = instruction.curr.execution_state.responsible_opcode()[0]
+    if isinstance(res_op, int):
+        res_op_int = int(res_op)
```
[the remark this answers is not in the document]
Comment (bchyl) [reply src/zkevm_specs/evm/execution/block_ctx.py:13]: yeah!

File: src/zkevm_specs/evm/execution/block_ctx.py (hunk 1 above)
Comment (bchyl) [reply src/zkevm_specs/evm/execution/block_ctx.py:14]: yeah, nice! when coding i also wonder for `responsible_opcode` and `responsible_opcode_lookup` have double checking for opcodeId.

File: src/zkevm_specs/util/param.py (hunk 2 above)
Comment (bchyl) [reply src/zkevm_specs/util/param.py:2]: last version testing typo, sorry.

File: src/zkevm_specs/evm/execution/block_ctx.py (hunk 3 above)
Comment (bchyl) [reply src/zkevm_specs/evm/execution/block_ctx.py:8]: yeah, in circuit code there is a  `blockctxGadget` for BlockContext ops.  `define a BlockContext execution state` is a good idea for reducing constraints. i will refactor it latter.

Thank you very much for your careful reviews.

File: specs/opcode/41COINBASE_45GASLIMIT_48BASEFEE.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (ed255) [reply specs/opcode/41COINBASE_45GASLIMIT_48BASEFEE.md]: > Maybe we can name it as `41COINBASE-45GASLIMIT_48BASEFEE.md`?

I understand that `41COINBASE_42TIMESTAMP_43NUMBER_44DIFFICULTY_45GASLIMIT_48BASEFEE.md` is a very long filename, although it makes it explicit which opcodes are defined in the file.  I'm OK with `41COINBASE-45GASLIMIT_48BASEFEE.md` then!

Note that if we move on with this https://github.com/appliedzkp/zkevm-specs/issues/154 we'll be able to add an index of all the opcodes, so the filenames will be less important :)

**Review by ed255 — APPROVED** [review]
LGTM! Thanks for working on this :)

**Review by han0110 — APPROVED** [review]
LGTM! Perhaps we can just have `BlockContextField::{Coinbase,Timestamp,Number,GasLimit,Difficulty,BaseFee}` identical to `opcode`, then we can use `opcode` as `tag` of lookup input. Also if #135 is adopted, then `ctx_expr` could be identical for all opcodes.

**Comment by barryWhiteHat** [comment]
@bchyl can you resolve conflicts here ? after that i think we can merge :)

**Pull request merged**